### ISSUE TITLE
Accuracy doesn't match

### ISSUE THREAD

**Issue by raghuramdr** [fchollet/keras#3849]
Please make sure that the boxes below are checked before you submit your issue. Thank you!
- [ ✓] Check that you are up-to-date with the master branch of Keras. You can update with:
  pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps
- [✓ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
  pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps
- [ ✓] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

The accuracy which is printed to the console in Keras doesn't match with the accuracy calculated from the confusion matrix. The code will make it more clear.

```
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Convolution2D, MaxPooling2D
from keras.layers import Activation, Dropout, Flatten, Dense
import numpy as np
import theano
from sklearn.metrics import classification_report, confusion_matrix
y_actual = np.ones((nb_test_samples),dtype = int)
y_actual[0:2817] = 0

train_datagen = ImageDataGenerator(
             featurewise_std_normalization=False,
             samplewise_std_normalization=False,
     rescale = 1./255)

test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
                    train_data_dir,
                    target_size = (img_width,img_height),
                    batch_size = 32,
                    class_mode = 'binary')

test_generator = test_datagen.flow_from_directory(
            test_data_dir,
                    target_size = (img_width,img_height),
                    batch_size = 32,
            class_mode = 'binary',
            shuffle = False         )

model.fit_generator(
                    train_generator,
                    samples_per_epoch = nb_train_samples,
                    nb_epoch = nb_epoch,
                    validation_data = test_generator,
                    nb_val_samples = nb_test_samples)

score =   model.evaluate_generator(
                      test_generator,
                      4938)
print "Test fraction correct (Accuracy) = {:.2f}".format(score[1])
prediction = model.predict_generator(test_generator,nb_test_samples)

for i in xrange(0,len(prediction)):
    if prediction[i]<0.5:
       prediction[i] = 0
    else:
       prediction[i] = 1

#y_predicted = test_generator.classes
print np.sum(prediction)
CM = confusion_matrix(y_actual,prediction)
print CM 
```

The accuracy printed onto the console shows around 91%, while that from the confusion matrix shows 47%. If I use y_predicted to calculate the confusion matrix, I get a confusion matrix which shows 100% accuracy, which is not matching with the accuracy printed on the console.  How do I resolve this issue?

**Comment by raghuramdr** [comment]
@vector-1127 I installed keras 1.0.5. It still doesn't work. From the Confusion matrix, I get an accuracy of 52%, and from the console, I am getting 90% accuracy. Is there anything wrong with the way I am calculating the confusion matrix?

**Comment by valentinap** [comment]
I'm having the same issue in R, how can I solve it?

**Comment by peymanrah** [comment]
Just a quick reply on this as I encountered the same issue and solved it by the code below:

test_datagen = ImageDataGenerator(
preprocessing_function=preprocess_input,rescale=1./255
)


validation_generator = test_datagen.flow_from_directory(

validation_dir,
batch_size=batch_size,
target_size = (width, height),
shuffle=False,#make sure shuffle is off or it is not gonna match!!
class_mode = "binary")

test_loss, test_acc = pre_trained_model.evaluate_generator(validation_generator, steps=round(nb_validation_samples/batch_size))
print('test acc:', test_acc) #acc:0.845
print('test loss:', test_loss) #0.58

predictions = pre_trained_model.predict_generator(validation_generator, steps=round(nb_validation_samples/batch_size))
test_preds = np.argmax(predictions, axis=-1)
l=test_preds.shape[0]
test_trues = validation_generator.classes
cm =confusion_matrix(test_trues[:l], test_preds)
print(cm) #acc:0.845 which matches with what I got from evaluate_generator above.
